velox/type/DecimalUtils.h

```diff
@@ -0,0 +1,122 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <string>
+#include "velox/common/base/Exceptions.h"
+#include "velox/type/LongDecimal.h"
+#include "velox/type/ShortDecimal.h"
+#include "velox/type/Type.h"
+
+namespace {
+
+std::string formatAsDecimal(
+    uint8_t scale,
+    facebook::velox::int128_t unscaledValue) {
+  if (unscaledValue == 0)
+    return "0";
+  std::string result;
+  if (unscaledValue < 0) {
+    result.append("-");
+    unscaledValue = ~unscaledValue + 1;
+  }
+  std::string unscaledStr = std::to_string(unscaledValue);
+  std::string formattedStr;
+  if (unscaledStr.length() <= scale) {
+    formattedStr.append("0");
+  } else {
+    formattedStr.append(unscaledStr.substr(0, unscaledStr.length() - scale));
+  }
+  if (scale > 0) {
+    formattedStr.append(".");
+    if (unscaledStr.length() < scale) {
+      for (auto i = 0; i < scale - unscaledStr.length(); ++i) {
+        formattedStr.append("0");
+      }
+      formattedStr.append(unscaledStr);
+    } else {
+      formattedStr.append(unscaledStr.substr(unscaledStr.length() - scale));
+    }
+  }
+  return result.append(formattedStr);
+}
+} // namespace
+
+namespace facebook::velox {
+
+static const int128_t POWERS_OF_TEN[]{
+    1,
+    10,
+    100,
+    1000,
+    10000,
+    100000,
+    1000000,
+    10000000,
+    100000000,
+    1000000000,
+    10000000000,
+    100000000000,
+    1000000000000,
+    10000000000000,
+    100000000000000,
+    1000000000000000,
+    10000000000000000,
+    100000000000000000,
+    1000000000000000000,
+    1000000000000000000 * (int128_t)10,
+    1000000000000000000 * (int128_t)100,
+    1000000000000000000 * (int128_t)1000,
+    1000000000000000000 * (int128_t)10000,
+    1000000000000000000 * (int128_t)100000,
+    1000000000000000000 * (int128_t)1000000,
+    1000000000000000000 * (int128_t)10000000,
+    1000000000000000000 * (int128_t)100000000,
+    1000000000000000000 * (int128_t)1000000000,
+    1000000000000000000 * (int128_t)10000000000,
+    1000000000000000000 * (int128_t)100000000000,
+    1000000000000000000 * (int128_t)1000000000000,
+    1000000000000000000 * (int128_t)10000000000000,
+    1000000000000000000 * (int128_t)100000000000000,
+    1000000000000000000 * (int128_t)1000000000000000,
+    1000000000000000000 * (int128_t)10000000000000000,
+    1000000000000000000 * (int128_t)100000000000000000,
+    1000000000000000000 * (int128_t)1000000000000000000,
+    1000000000000000000 * (int128_t)1000000000000000000 * (int128_t)10,
+    1000000000000000000 * (int128_t)1000000000000000000 * (int128_t)100};
+
+template <typename T>
+inline std::string decimalToString(const T& value, const TypePtr& type) {
+  VELOX_UNSUPPORTED();
+}
+
+template <>
+inline std::string decimalToString<LongDecimal>(
```

Comment:
declare these here and define them in a cpp file. This will avoid recompiling these functions for every use.

Reply:
and these do not have to be inlined.